File: lang/c/src/vpkppc/format/SDAT.cpp
#include <vpkppc/format/SDAT.h>

#include <sourceppc/Helpers.h>

using namespace sourceppc;
using namespace vpkpp;

const uint32_t VPKPP_SDAT_SIGNATURE = SDAT_SIGNATURE;
const char* VPKPP_SDAT_EXTENSION    = SDAT_EXTENSION.data();

VPKPP_API vpkpp_pack_file_handle_t vpkpp_sdat_create(const char* path) {
	SOURCEPP_EARLY_RETURN_VAL(path, nullptr);

	auto packFile = SDAT::create(path);
	if (!packFile) {
		return nullptr;
	}
	return packFile.release();
}

VPKPP_API vpkpp_pack_file_handle_t vpkpp_sdat_open(const char* path, vpkpp_entry_callback_t callback) {
	SOURCEPP_EARLY_RETURN_VAL(path, nullptr);

	auto packFile = SDAT::open(path, callback ? [callback](const std::string& entryPath, const Entry& entry) {
		callback(entryPath.c_str(), const_cast<Entry*>(&entry));
	} : static_cast<PackFile::EntryCallback>(nullptr));
	if (!packFile) {
		return nullptr;
	}
	return packFile.release();
}

File: lang/csharp/src/vpkpp/Format/SDAT.cs
using System;

namespace sourcepp.vpkpp.Format;

using OpenPropertyRequest = Func<PackFile, OpenProperty, byte[]>;

using EntryCallback = Action<string, Entry>;

public class SDAT : PackFile
{
	protected SDAT(nint handle, bool managed = true) : base(handle, managed)
	{
	}

	public static SDAT? Create(string path)
	{
		var handle = DLL.vpkpp_sdat_create(path);
		return handle == nint.Zero ? null : new SDAT(handle);
	}

	public new static SDAT? Open(string path, EntryCallback? callback = null, OpenPropertyRequest? _ = null)
	{
		var handle = DLL.vpkpp_sdat_open(path, callback is not null ? (entryPath, entry) =>
		{
			callback(entryPath, new Entry(entry, false));
		} : null);
		return handle == nint.Zero ? null : new SDAT(handle);
	}
}
